File: src/main/java/com/naminhyeok/fantazzk/template/TemplateConfiguration.java
package com.naminhyeok.fantazzk.template;

import org.jmolecules.ddd.types.ValueObject;

public record TemplateConfiguration(
    TemplateMode mode,
    int teamCount,
    int teamSize,
    Integer budget,
    DraftOrderStrategy draftOrderStrategy
) implements ValueObject {
    public TemplateConfiguration {
        if (teamCount <= 0) {
            throw new IllegalArgumentException("팀 수는 0보다 커야 합니다");
        }

        if (teamSize <= 0) {
            throw new IllegalArgumentException("팀 크기는 0보다 커야 합니다");
        }

        if (mode == TemplateMode.AUCTION) {
            if (budget == null) {
                throw new IllegalArgumentException("경매 템플릿에는 예산이 필요합니다");
            }
            if (budget <= 0) {
                throw new IllegalArgumentException("예산은 0보다 커야 합니다");
            }
            if (draftOrderStrategy != null) {
                throw new IllegalArgumentException("경매 템플릿에는 드래프트 순서 전략을 지정할 수 없습니다");
            }
        }

        if (mode == TemplateMode.DRAFT) {
            if (budget != null) {
                throw new IllegalArgumentException("드래프트 템플릿에는 예산을 지정할 수 없습니다");
            }
            if (draftOrderStrategy == null) {
                throw new IllegalArgumentException("드래프트 템플릿에는 순서 전략이 필요합니다");
            }
        }
    }

    public static TemplateConfiguration auction(int teamCount, int teamSize, int budget) {
        return new TemplateConfiguration(TemplateMode.AUCTION, teamCount, teamSize, budget, null);
    }

    public static TemplateConfiguration draft(int teamCount, int teamSize, DraftOrderStrategy strategy) {
        return new TemplateConfiguration(TemplateMode.DRAFT, teamCount, teamSize, null, strategy);
    }

    public static TemplateConfiguration from(
        TemplateMode mode,
        int teamCount,
        int teamSize,
        Integer budget,
        DraftOrderStrategy draftOrderStrategy
    ) {
        return switch (mode) {
            case AUCTION -> {
                if (draftOrderStrategy != null) {
                    throw new IllegalArgumentException("경매 템플릿에는 드래프트 순서 전략을 지정할 수 없습니다");
                }
                yield auction(teamCount, teamSize, requireBudget(budget));
            }
            case DRAFT -> {
                if (budget != null) {
                    throw new IllegalArgumentException("드래프트 템플릿에는 예산을 지정할 수 없습니다");
                }
                yield draft(teamCount, teamSize, requireStrategy(draftOrderStrategy));
            }
        };
    }

    public int requiredPlayerCount() {
        return teamCount * (teamSize - 1);
    }

    private static int requireBudget(Integer budget) {
        if (budget == null) {
            throw new IllegalArgumentException("경매 템플릿에는 예산이 필요합니다");
        }
        return budget;
    }

    private static DraftOrderStrategy requireStrategy(DraftOrderStrategy strategy) {
        if (strategy == null) {
            throw new IllegalArgumentException("드래프트 템플릿에는 순서 전략이 필요합니다");
        }
        return strategy;
    }
}

File: template/model/src/main/kotlin/com/naminhyeok/fantazzk/template/TemplateConfiguration.kt
package com.naminhyeok.fantazzk.template

sealed interface TemplateConfiguration {
    val mode: TeamBuildingMode
    val teamCount: Int
    val teamSize: Int
    val budget: Int?
    val draftOrderStrategy: DraftOrderStrategy?
    val requiredPlayerCount: Int

    data class Auction(
        override val teamCount: Int,
        override val teamSize: Int,
        val budgetValue: Int,
    ) : TemplateConfiguration {
        init {
            require(teamCount > 0) { "팀 수는 0보다 커야 합니다" }
            require(teamSize > 0) { "팀 크기는 0보다 커야 합니다" }
            require(budgetValue > 0) { "예산은 0보다 커야 합니다" }
        }

        override val mode: TeamBuildingMode = TeamBuildingMode.AUCTION
        override val budget: Int = budgetValue
        override val draftOrderStrategy: DraftOrderStrategy? = null
        override val requiredPlayerCount: Int = teamCount * (teamSize - 1)
    }

    data class Draft(
        override val teamCount: Int,
        override val teamSize: Int,
        val strategy: DraftOrderStrategy,
    ) : TemplateConfiguration {
        init {
            require(teamCount > 0) { "팀 수는 0보다 커야 합니다" }
            require(teamSize > 0) { "팀 크기는 0보다 커야 합니다" }
        }

        override val mode: TeamBuildingMode = TeamBuildingMode.DRAFT
        override val budget: Int? = null
        override val draftOrderStrategy: DraftOrderStrategy = strategy
        override val requiredPlayerCount: Int = teamCount * (teamSize - 1)
    }

    companion object {
        fun from(
            mode: TeamBuildingMode,
            teamCount: Int,
            teamSize: Int,
            budget: Int?,
            draftOrderStrategy: DraftOrderStrategy?,
        ): TemplateConfiguration =
            when (mode) {
                TeamBuildingMode.AUCTION -> {
                    require(draftOrderStrategy == null) { "경매 템플릿에는 드래프트 순서 전략을 지정할 수 없습니다" }
                    Auction(
                        teamCount = teamCount,
                        teamSize = teamSize,
                        budgetValue = requireNotNull(budget) { "경매 템플릿에는 예산이 필요합니다" },
                    )
                }

                TeamBuildingMode.DRAFT -> {
                    require(budget == null) { "드래프트 템플릿에는 예산을 지정할 수 없습니다" }
                    Draft(
                        teamCount = teamCount,
                        teamSize = teamSize,
                        strategy = requireNotNull(draftOrderStrategy) { "드래프트 템플릿에는 순서 전략이 필요합니다" },
                    )
                }
            }
    }
}
